src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -5794,13 +5799,27 @@ and TcExprLazy (cenv: cenv) overallTy env tpenv (synInnerExpr, m) =
     let expr = mkLazyDelayed g m innerTy (mkUnitDelayLambda g m innerExpr)
     expr, tpenv
 
+and CheckTupleIsCorrectLength g (env: TcEnv) m tupleTy args typeCheckArgs =
```

Comment:
@dsyme since the same code is needed in `CheckPatterns.TcPatTuple` I factored it out into this function. However it does require to pass in the type checking function. Is this ok, or should we just duplicate the code in the 2 places? Or is there a third and better option? :)

Reply:
In general we don't do this second-order factoring in the compiler (i.e. we don't pass around functions that operate over the core env/cenv objects in different phases).  There are some exceptions.  I don't really mind adding a few but in general I think we should be more focusing on much simpler first-order factoring of common code fragments.   Second-order factoring just makes code harder to read and understand, even if it shares some common code.

Reply:
(The factoring you have is fine)